### DISCUSSION TITLE
How to connect dosemu2 serial port to socat to Linux physical serial port?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1738] [General]
I'm currently using a setup with two dosemu2 instances. They both connect to one end each of a socat instance called like so: `socat -r log.txt pty,link=/tmp/vptty-debug,rawer pty,link=/tmp/vptty-terminal,rawer`

The terminal end is connected to a dosemu2 run like this: `dosemu -K "$PWD" -E "c:ldebug.com /c=g terminal.com a 4 ! &" -I "serial { com 1 device /tmp/vptty-terminal }" -I "serial { com 2 device /tmp/vptty-dos }"` (Terminal's `a` means COM1, `4` means 9600 baud, `!` means skip slideshow intro, `&` means default to send-CR-only policy for Enter keypresses. This [is Eric Auer's Terminal](https://hg.pushbx.org/ecm/terminal) with some mods by me, a GNU GPL v2+ DOS terminal application.)

The debugger end is connected to another dosemu2, run like so: `dosemu -dumb -td -kt -I "serial { com 2 device /tmp/vptty-debug }" -E "ldebug.com /c='r dco or= 4800'"` (lDebug defaults to COM2, 9600 baud.)

Further, I can use the `log.txt` written by the socat as input into my tractest and TracList. This works reasonably well. (As mentioned on the FreeDOS mailing lists, long output from the debugger to the terminal sometimes results in losing parts of the output. Probably not a dosemu-specific problem.)

Now, I'd like to connect the Terminal dosemu to the physical serial port of this Linux box, which connects with a null-modem cable to COM2 on a DOS desktop computer next to it. I found that I can make this work by directly connecting dosemu to the /dev/ttyS0 device file. This goes like `dosemu -K "$PWD" -E "c:ldebug.com /c=g terminal.com a 4 ! &" -I "serial { com 1 device /dev/ttyS0 }" -I "serial { com 2 device /tmp/vptty-dos }"` and then I can run `ldebug /c='r dco or= 4000'` on the DOS machine and the terminal picks it up correctly.

But I also want to have the option given by `socat` to dump the binary data sent from the DOS machine to the Linux machine into a log file. I have not yet managed to find a way to do this. My first attempt was using `socat`, but I can't figure out the correct way to do this. This is the command I use: `sudo socat /dev/ttyS0,rawer,b9600,nonblock pty,link=/tmp/vptty-arona,rawer,mode=666` and dosemu2 run as `dosemu -K "$PWD" -E "c:ldebug.com /c=g terminal.com a 4 ! &" -I "serial { com 1 device /tmp/vptty-arona }" -I "serial { com 2 device /tmp/vptty-dos }"`. Both run without error messages but there is no connection.

Next I tried just connecting dosemu directly to /dev/ttyS0 as before, but to then run `cat < /dev/ttyS0`. This does output the responses by the debugger to the stdout of the `cat` process. But it doesn't pass through the same output to the Terminal DOS process any longer.

Any ideas how to do this?

**Comment by ecm-pushbx** [comment]
Also tried use case 1 of https://github.com/danielinux/ttybus but that just connects the `cat < /tmp/ttyS0fake0` to the Terminal connected to `/tmp/ttyS0fake1` and doesn't seem to care about the `/dev/ttyS0` at all.

**Comment by stsp** [comment]
> My first attempt was using socat, but I can't figure out the correct way to do this.

Obviously you need to use the
`-r` and `-R` options of socat.

**Reply by ecm-pushbx** [reply]
> > My first attempt was using socat, but I can't figure out the correct way to do this.
> 
> Obviously you need to use the `-r` and `-R` options of socat.

I know that. As I already stated in the dosemu-to-dosemu example the `-r` option works fine. However, that won't help connectivity when the dosemu-to-physical-DOS-machine I/O doesn't work.

**Reply by ecm-pushbx** [reply]
To clarify, the dosemu-to-physical test doesn't show any of the debugger output on the dosemu running Eric's Terminal and (presumably) doesn't send any of the Terminal input to the debugger.

**Reply by stsp** [reply]
Ah so you stick socat in between
and things stop working?
Note that things like RTS, DTR
are not passed via socat. So you
need to disable HW flow control
in that case.
But this is likely not what you want,
so maybe you want the logging
on dosemu side instead.
IIRC some higher debug levels
for serial port dump entire stream
into the log. From which you can
grep. Or if currently not dumped,
this can be trivially added (by yourself :)

**Reply by ecm-pushbx** [reply]
> Ah so you stick socat in between and things stop working?

Yes, exactly.

> Note that things like RTS, DTR are not passed via socat. So you need to disable HW flow control in that case. But this is likely not what you want,

My theory was that connecting two dosemus works fine, so the socat between a dosemu and a physical port should also work. I don't think [lDebug's serial I/O](https://hg.pushbx.org/ecm/ldebug/file/77d4681a4028/source/serialp.asm) uses hardware flow control. Terminal has options for that but I think the default is also to not use any. (I also use a program called Public Domain ZModem on the DOS side and that most likely uses hardware flow control, but I use Linux native `rz` and `sz` to talk to that, not a dosemu. And that works just fine.)

> so maybe you want the logging on dosemu side instead. IIRC some higher debug levels for serial port dump entire stream into the log. From which you can grep. Or if currently not dumped, this can be trivially added (by yourself :)

Might do that or might add logging to Terminal itself (to a DOS file handle then). I'd much prefer leveraging existing tools like socat though.

**Reply by stsp** [reply]
Well, maybe you are hitting some
bug in dosemu2. The only way to check
is to compare the log files between
working and socat-broken sessions.
Also if you log things with socat, then
you can figure out which side didn't
send or receive something...

**Reply by ecm-pushbx** [reply]
Running `$ sudo socat -v /dev/ttyS0,rawer,b9600,nonblock pty,link=/tmp/vptty-arona,rawer,mode=666; echo` I get the following output:

```
< 2022/07/20 19:11:32.781769  length=1 from=0 to=0
t< 2022/07/20 19:11:33.166266  length=1 from=1 to=1
e< 2022/07/20 19:11:33.477358  length=1 from=2 to=2
s< 2022/07/20 19:11:33.806871  length=1 from=3 to=3
t
```

This is the input "test" entered into dosemu's Terminal. So that part works. However, no data is logged here from the DOS box's serial port to socat.

This is the good log: https://pushbx.org/ecm/test/20220720/good.log You can read that dosemu receives the prompt from the debugger, sends "keep" and Enter and "q", and receives the line editing output. This is the command used: `$ ~/local.new/bin/dosemu -K "$PWD" -E "c:ldebug.com /c=g terminal.com a 4 ! &" -I "serial { com 1 device /dev/ttyS0 }" -I "serial { com 2 device /tmp/vptty-dos }" -D+9s`

This is the bad log: https://pushbx.org/ecm/test/20220720/bad.log Nothing is received, "test" is sent. The command used was `$ ~/local.new/bin/dosemu -K "$PWD" -E "c:ldebug.com /c=g terminal.com a 4 ! &" -I "serial { com 1 device /tmp/vptty-arona }" -I "serial { com 2 device /tmp/vptty-dos }" -D+9s`

So given the socat log I suspect this is not a dosemu error.

**Reply by stsp** [reply]
Maybe you need to use
`gopen:` or `open:` before ttyS0?

**Comment by ecm-pushbx** [comment]
Thanks for the suggestions, no change however. I did figure out that the input to Eric's Terminal does actually arrive at the debugger. I can press Enter to end the keep prompt, I can even enter "keep" then Enter and then regular commands (like Q) to control the debugger. However, socat and dosemu both never receive any of the debugger's output.

**Comment by ecm-pushbx** [comment]
Okay! I did get it to work finally. Not sure why but the answer to https://unix.stackexchange.com/questions/378400/socat-uart-logging-and-redirecting with that exact command (except `/dev/ttyS0` instead of USB) works, whereas mine didn't work. Still have to figure out exactly what part went wrong. Thanks for the help.

**Reply by stsp** [reply]
Different speed settings?
Yes, indeed... with socat, dosemu
can't set the baudrate.

**Reply by ecm-pushbx** [reply]
Oh, no, I also changed the baud rate to 9600 to match my debugger application. Testing so far seems like `rawer` on the `/dev/ttyS0` is the problem, just `raw` works fine.

**Reply by ecm-pushbx** [reply]
Oddly enough, `raw,echo=0` also works, but according to the manual page `rawer` is preferred over the "obsolete" `raw`:

```
       raw    Sets raw mode, thus passing input and output almost unprocessed.
              This option is obsolete, use option rawer or cfmakeraw instead.

       rawer  Makes terminal rawer than raw  option.  This  option  implicitly
              turns off echo. (example).
```

**Reply by stsp** [reply]
Hmmm. :)

**Reply by stsp** [reply]
cfmakeraw is also to try.

**Reply by ecm-pushbx** [reply]
`cfmakeraw` without or with the `echo=0` also works fine.

**Reply by stsp** [reply]
So you seem to have found a
bug in socat, which is that rawer
doesn't disable echo, contrary to
what is advertised?

**Reply by ecm-pushbx** [reply]
That's not the bug, echo doesn't seem to matter. It's just unclear *what* exactly socat's `rawer` does, but it differs in some way from `raw` and `cfmakeraw` (both either with or without `echo=0`). And that isn't documented well enough and no one knows to tell you to look for it.

**Reply by stsp** [reply]
Are you sure echo doesn't seem
to matter? You mean, with echo=1
it would still work?

**Reply by ecm-pushbx** [reply]
Anyway problems that I had earlier today with PD ZModem are also caused by connecting socat with `rawer` to the port. ZM would say it was "Waiting for sender" indefinitely, while I tried to send using rz from Linux. Now when I sent a Control-C to the *sender* side (Linux rz), the *receiver* side (DOS PDZM) would detect a "Transfer aborted" immediately and close itself with that. So very likely, whatever `rawer` did, it persisted and caused problems even after socat was already gone.

When I had this zmodem problem earlier, I powercycled the DOS box, suspended and resumed the Linux box, to no avail. The error condition was finally solved by connecting lDebug from DOS to the Linux side dosemu running Terminal (with `-I "serial { com 1 device /dev/ttyS0 }"` for dosemu).

**Reply by stsp** [reply]
https://stackoverflow.com/questions/69634907/cant-read-from-serial-port-with-socat-and-rawer-option
Here they say rawer breaks
the speed setting.

**Reply by ecm-pushbx** [reply]
> Are you sure echo doesn't seem to matter? You mean, with echo=1 it would still work?

No, it doesn't. I was referring only to no echo setting or echo=0. But with raw,echo=1 I get garbled input and output on both ends, and this is a very different error to what rawer caused.

**Reply by ecm-pushbx** [reply]
> https://stackoverflow.com/questions/69634907/cant-read-from-serial-port-with-socat-and-rawer-option Here they say rawer breaks the speed setting.

Might be related or the same problem. Or, one of the problems, perhaps. Thanks for unearthing this.

**Reply by stsp** [reply]
```
   case OPT_RAW:
      _xiotermios_data.termarg.c_iflag &=
         ~(IGNBRK|BRKINT|IGNPAR|PARMRK|INPCK|ISTRIP|INLCR|IGNCR|ICRNL|IXON|IXOFF
|IUCLC|IXANY|IMAXBEL);
      _xiotermios_data.termarg.c_iflag |= (0);
      _xiotermios_data.termarg.c_oflag &= ~(OPOST);
      _xiotermios_data.termarg.c_oflag |= (0);
      _xiotermios_data.termarg.c_cflag &= ~(0);
      _xiotermios_data.termarg.c_cflag |= (0);
      _xiotermios_data.termarg.c_lflag &= ~(ISIG|ICANON|XCASE);
      _xiotermios_data.termarg.c_lflag |= (0);
      _xiotermios_data.termarg.c_cc[VMIN] = 1;
      _xiotermios_data.termarg.c_cc[VTIME] = 0;
      break;
   case OPT_TERMIOS_RAWER:
      _xiotermios_data.termarg.c_iflag = 0;
      _xiotermios_data.termarg.c_oflag = 0;
      _xiotermios_data.termarg.c_lflag = 0;
      _xiotermios_data.termarg.c_cflag = (CS8);
      _xiotermios_data.termarg.c_cc[VMIN] = 1;
      _xiotermios_data.termarg.c_cc[VTIME] = 0;
      break;
```

This is clearly wrong.
I think this rawer thing is all but safe,
and in your case this may help:
```
-       _xiotermios_data.termarg.c_cflag = (CS8);
+      _xiotermios_data.termarg.c_cflag = (CS8) | (_xiotermios_data.termarg.c_cflag & CBAUD);
```

Would you like to try?

**Reply by ecm-pushbx** [reply]
Only if you walk me through how to build this thing, I don't really want to spend any more time on this.

**Reply by stsp** [reply]
I did `configure` and then `make -j 9`
and it built w/o any problems.

**Reply by ecm-pushbx** [reply]
Downloaded from http://www.dest-unreach.org/socat/download/socat-1.7.4.3.tar.gz via http://www.dest-unreach.org/socat/ and tested before patching and with your patch. Both behave exactly the same as the system's preinstalled socat.

**Reply by stsp** [reply]
So it zeroes iflags, oflags, lflags, cflags.
Would you like to check which one leads
to a problem, by replacing the appropriate
zero-out with the handling from OPT_RAW
block?

**Reply by ecm-pushbx** [reply]
I have worked enough on it today, maybe another day. Thanks for the help in any case.

**Reply by stsp** [reply]
I understand, thanks.
Just note that socat probably has
more users than dosemu2+freedos
combined, so fixing it kinda makes
sense. :)
